src/main/java/io/anserini/rerank/AxiomReranker.java

```
@@ -0,0 +1,302 @@
+  private Set<Integer> selectDocs(ScoredDocuments docs, RerankerContext context)
+      throws IOException {
+    Set<Integer> docidSet = new HashSet<>(Arrays.asList(ArrayUtils.toObject(
+        Arrays.copyOfRange(docs.ids, 0, Math.min(this.M, docs.ids.length)))));
+    long targetSize = this.R * this.M;
+
+    if (docidSet.size() < targetSize) {
+      IndexSearcher searcher = context.getIndexSearcher();
+      IndexReader reader = searcher.getIndexReader();
+      int availableDocsCnt = reader.getDocCount(this.field);
+      List<Integer> docidsPoolList = new ArrayList<>();
+      for (int i = 1; i <= availableDocsCnt; i++) {
+        docidsPoolList.add(i);
+      }
+      Random random = new Random();
+      while (docidSet.size() < targetSize) {
+        docidSet.add(docidsPoolList.get(random.nextInt(docidsPoolList.size())));
+      }
+    }
+
+    return docidSet;
+  }
+
+  /** Extract ALL the terms from the documents pool.
+   *
+   *  @param docIds          The reranking pool, see {@link #selectDocs} for explanations
+   *  @param context         An instance of RerankerContext
+   *  @param filterPattern   A Regex pattern that terms are collected only they matches the pattern, could be null
+   *
+   *  @return A Map of <term -> Set<docId>> kind of a small inverted list where the Set of docIds is where the term occurs
+   */
+  @VisibleForTesting
+  private Map<String, Set<Integer>> extractTerms(Set<Integer> docIds, RerankerContext context,
+                                                 Pattern filterPattern) throws Exception, IOException {
+    IndexSearcher searcher = context.getIndexSearcher();
+    IndexReader reader = searcher.getIndexReader();
+    Map<String, Set<Integer>> termDocidSets = new HashMap<>();
+    for (int docid : docIds) {
+      Terms terms = reader.getTermVector(docid, LuceneDocumentGenerator.FIELD_BODY);
+      if (terms == null) {
+        LOG.info("Document vector not stored for docid: " + docid);
+        continue;
```

Do we actually want to throw an exception here? If doc vector not stored, we can't really do query expansion...

---

No...
When I randomly pick documents from the whole index I used an array [1, maxDocId]. What I found is that some docIds are missing from the consecutive integer array. My guess is that when indexing the document there might be problems, e.g. empty doc, that leave the document not storing the termVector.